index.js

```diff
@@ -298,68 +324,88 @@ function cmpType (a, b) {
   return typeOrder[a.type] - typeOrder[b.type]
 }
 
-DB.prototype._onpt = function (pt, seen, cb) {
+// Given a node by its version, this collects the node itself, and also
+// recursively climbs all ways and relations that the node (or its referers)
+// are referred to by.
+// OsmVersion, { OsmVersion: Boolean } -> [OsmDocument]
+DB.prototype._collectNodeAndReferers = function (version, seenAccum, cb) {
   cb = once(cb || noop)
   var self = this
-  var link = pt.value.toString('hex')
-  if (has(seen, link)) return cb(null, [])
-  seen[link] = true
+  if (has(seenAccum, version)) return cb(null, [])
+  seenAccum[version] = true
   var res = [], added = {}, pending = 2
-  self.log.get(link, function (err, doc) {
-    if (doc && doc.value && doc.value.k && doc.value.v) {
-      addDoc(doc.value.k, link, doc.value.v)
-    }
+
+  self.log.get(version, function (err, node) {
+    addDocFromNode(node)
     if (--pending === 0) cb(null, res)
   })
-  self._links(link, function onlinks (err, links) {
+
+  self._getReferers(version, function onlinks (err, links) {
     if (!links) links = []
     links.forEach(function (link) {
-      if (has(seen, link)) return
-      seen[link] = true
+      if (has(seenAccum, link)) return
+      seenAccum[link] = true
       pending++
-      self.log.get(link, function (err, doc) {
-        if (doc && doc.value && doc.value.k && doc.value.v) {
+      self.log.get(link, function (err, node) {
+        addDocFromNode(node)
+        if (node && node.value && node.value.k && node.value.v) {
           pending++
-          self.get(doc.value.k, function (err, xdocs) {
+          self.get(node.value.k, function (err, docs) {
             if (err) return cb(err)
-            Object.keys(xdocs).forEach(function (key) {
-              addDoc(doc.value.k, key, xdocs[key])
+            Object.keys(docs).forEach(function (key) {
+              addDoc(node.value.k, key, docs[key])
             })
             if (--pending === 0) cb(null, res)
           })
         }
-        addDoc(doc.value.k, link, doc.value.v)
         if (--pending === 0) cb(null, res)
       })
       pending++
-      self._links(link, function (err, links) {
+      self._getReferers(link, function (err, links) {
         onlinks(err, links)
```

Comment:
de-inlining the `onlinks` function would clarify that `links` here is referring to `links` inside this closure, not the `links` above.